### ISSUE TITLE
[DEV-2.0] Anyone tried OpenId?

### ISSUE THREAD

**Issue by WimDH** [zalando/connexion#694]
### Description
I tried to use OpenId for secure my endpoints, but when the security is not applied. My endpoint returned it's result without the use of my token.

I cloned the dev-2.0 branch and used the hello example.

My swagger config file:

```
openapi: "3.0.0"

info:
  title: Hello World
  version: "1.0"
servers:
  - url: http://localhost:9090/v1.0

paths:
  /greeting/{name}:
    post:
      summary: Generate greeting
      description: Generates a greeting message.
      operationId: hello.post_greeting
      responses:
        200:
          description: greeting response
          content:
            text/plain:
              schema:
                type: string
                example: "hello dave!"
      parameters:
        - name: name
          in: path
          description: Name of the person to greet.
          required: true
          schema:
            type: string
            example: "dave"

components:
  securitySchemes:
    openId:
      type: openIdConnect
      openIdConnectUrl: http://127.0.0.1:8080/.well-known/openid-configuration

# Global security
security:
  - openId:
      - admin
```

`curl http://127.0.0.1:9090/v1.0/greeting/wim --request POST`

produces:
`hello wim`

**Comment by hjacobs** [comment]
I think OpenID is not implemented, feel free to propose an implementation/PR (?).

**Comment by dtkav** [comment]
Yeah, sorry @WimDH I haven't had a chance to fully understand OpenID Connect and implement it.
I've been trying to focus on getting the OpenAPI 3 support released with feature parity to Swagger 2, and then after that's released I'm happy to start adding features like this. This is definitely high on the list.

**Comment by cziebuhr** [comment]
You can still use oauth2 and parse the JWT yourself in x-tokenInfoFunc. Currently there's no programmatic way to access the current token from the authorization header (apart from parsing it again). Maybe I could add that to #686.

**Comment by styk-tv** [comment]
@hjacobs @dtkav @cziebuhr from what I understand:

One of `.well-known/openid-configuration`-listed endpoints is called `certs`.  It carries a list of dictionaries that are compliant with JWK (Json Web Key RFC 7517). In my Keycloak there is only one and it carries enough info to reconstruct a public key that is required to verify a signature of a JWT token.

in python `RSAAlgorithm.from_jwk(keys[0])` where `RSAAlgorithm` comes from `jwt` package and `keys` is a direct output of `certs` endpoint.

Once signature on token is verified to be coming from the `.well-known` server, there are usually other checks like `aud`ience, expiry and so on.

As an an example unrelated directly to openapi specification i will list below kubernetes spec to demonstrate the difference in groups handling:

```
spec:
  kubeAPIServer:
    oidcIssuerURL: https://your-oidc-provider.svc.cluster.local
    oidcClientID: kubernetes
    oidcUsernameClaim: email
    oidcGroupsClaim: groups  
```

In JWT token:
<img width="301" alt="screenshot 2018-12-24 at 06 54 19" src="https://user-images.githubusercontent.com/838921/50392741-cb381200-0748-11e9-91bd-1db469afa319.png">

might then very easily translate to in openapi.yaml:
```
paths:
  /pets/{petId}:
    get:
      summary: Get a pet by ID
      security:
        - openId:
          - api-user
```

Reason I'm listing this is I believe it is important to observe implementation that cares where the GroupClaims are coming from. As you are crafting the token, groups can be placed in different locations. I use `groups`, but as you are defining a mapping it could be any name. I have not seen such placeholder in the OpenApi3 Discovery specification. I would kindly ask you consider this during implementation.

In Summary: OpenID Connect Discover (oidc) is very close from being fully realized in python for OpenApi. Personally I have both beautiful dreams and nightmares about this section for a long time. I think for all of us, getting to a point where you can use abstracted system like Keycloak, Auth0, Google, Cognito ... with or without federated LDAP that admit control to selected endpoints based on token's self-contained details is very long awaited.

**Comment by rubenfanjulestrada** [comment]
Hello, I am just curious if there is any improvement in this issue.

**Comment by rfanjul** [comment]
Hello I am just curious if there was an improvement in this issue. It is something I would like to use

**Comment by julien10** [comment]
Yep, same for me, are there any news ?

**Comment by gbmarc1** [comment]
Still no news on this issue?

**Comment by sindre-nistad** [comment]
This feature would be much appreciated.

**Comment by rhernaus** [comment]
I'm curious if there is any news on this issue. I would really like to use this.

**Comment by georgms** [comment]
+1

**Comment by sknopp** [comment]
+1

**Comment by dmarkey** [comment]
This would be great, oidc is becoming de-facto.

**Comment by dmarkey** [comment]
From what I can tell, it could be as simple as this:
https://github.com/dmarkey/connexion/pull/1/files

But leaves it up to the application to do the JWT validation.

**Comment by chrisinmtown** [comment]
I am using an OpenID Connect server to guard my website as follows. The site uses Apache Httpd as a front-end. The front-end proxies requests to a Connexion back-end REST API server. I configured Httpd with the module `mod_auth_openidc` which handles redirecting the user's browser to the appropriate URL to authenticate, multi-factor, yada yada. On each request to the back-end, Httpd includes the user name claim from the OpenID Connect server, and the Connexion app security function checks whether the user has sufficient privileges.

I'm not quite certain whether this answers the original question. The question might be, can you configure a Connexion app to do *all* of the OpenID Connect authentication flow. I don't have an answer for that.
